sweep/txgenerator.go

```diff
@@ -96,9 +97,40 @@ func (s *UtxoSweeper) generateInputPartitionings(sweepableInputs []txInput,
 			return sets, nil
 		}
 
+		// We fetch any UTXOs already used by inputs of the same
+		// exclusive groups, since we prefer to re-use them instead of
+		// adding new wallet inputs.
+		utxos := make(map[wire.OutPoint]*lnwallet.Utxo)
+		exclStr := ""
+		first := true
+		for eg := range txInputs.exclusiveGroups {
+			for _, utxo := range s.exclusiveGroupUtxos[eg] {
+				utxos[utxo.OutPoint] = utxo
+			}
+
+			if first {
+				exclStr += fmt.Sprintf("%d", eg)
+			} else {
+				exclStr += fmt.Sprintf(", %d", eg)
+			}
+			first = false
+		}
+		exclStr += ""
+
+		reuseUtxos := make([]*lnwallet.Utxo, 0, len(utxos))
+		for _, utxo := range utxos {
+			reuseUtxos = append(reuseUtxos, utxo)
+		}
+
+		if len(reuseUtxos) > 0 {
+			log.Infof("Re-using %d UTXOs for exclusive groups [%s]",
+				len(reuseUtxos), exclStr)
+		}
+
 		// Check the current output value and add wallet utxos if
 		// needed to push the output value to the lower limit.
-		if err := txInputs.tryAddWalletInputsIfNeeded(); err != nil {
+		err := txInputs.tryAddWalletInputsIfNeeded(reuseUtxos)
```

Comment:
should perhaps consider removing the saved UTXOs after a while, or checking whether they get spent by some other tx that is not in the same exclusive group. Otherwise maybe we can end up, trying to spend from a spent output again and again?

Reply:
Yeah seems this could cause us to go into a spend-then-remove loop, that eventually fails the input all together. However, as long as the other sub-systems are doing proper UTXO locking, I think this'll be rare in practice, given that the sweeper is the only system that will purposefully double spend.

Reply:
if we restart the node, now does it recreate the exclusive groups that were previously assigned?

Reply:
The UXTOs assigned to each exclusive group should always be different after a restart assuming a previous successful sweep broadcast happened (will always happen with Neutrino due to lack of mempool) since they will no longer be available for coin selection. This is unavoidable unless we introduce some state into the sweeper.

> should perhaps consider removing the saved UTXOs after a while, or checking whether they get spent by some other tx that is not in the same exclusive group.

They could also be spent by other means that use manual coin selection with output leases since the coin selection done in the sweeper does not account for that. I think this, in combination with checking for spentness of wallet UTXOs, may be required to ensure we can eventually arrive at a successful sweep.